File: smartcore-wasi/node/perf.js
const fs = require("fs");
const wasiImport = require("wasi");
const { performance } = require("perf_hooks");
const { WASI } = wasiImport;
const os = require("os");

const getStringPointer = (data, instance) => {
  var enc = new TextEncoder(); // always utf-8
  var length = data.length + 1;
  const targetMem = new Uint8Array(length);
  enc.encodeInto(data, targetMem);
  targetMem[length - 1] = 0;

  var ptr = instance.exports.allocate(length);
  var mem = new Uint8Array(instance.exports.memory.buffer, ptr, length);
  mem.set(targetMem);
  return { ptr, len: length };
};

const wasi = new WASI({
  args: [],
  env: process.env,
  preopens: {
    ".": ".",
  },
});

const importObject = { wasi_snapshot_preview1: wasi.wasiImport };
const relativeDir = "lib/js";

let instance;

const initialize = async () => {
  console.log("Initializing module");
  const wasm = await WebAssembly.compile(
    fs.readFileSync(`${relativeDir}/smartcore_wasi_lib.wasm`)
  );
  instance = await WebAssembly.instantiate(wasm, importObject);
  wasi.initialize(instance);
  let { ptr: fileToCopyPtr, len: len1 } = getStringPointer(
    `${relativeDir}/iris_knn.model`,
    instance
  );
  instance.exports.init(fileToCopyPtr);
  console.log("Initialization complete");
};

initialize().then(() => {
  const performances = [];
  let startTime = new Date();

  for (let i = 0; i < 1000; i++) {
    const t0 = process.hrtime.bigint();
    instance.exports.load_model();
    const t1 = process.hrtime.bigint();
    performances.push(t1 - t0);
  }
  const endTime = new Date();
  const content = `startTime:${startTime.toUTCString()}\nendTime:${endTime.toUTCString()}\ndata:[${performances}]`;
  require("fs").writeFileSync("data-js.csv", content);
});

File: smartcore-wasi/go/perf.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"time"

	wasmer "github.com/wasmerio/wasmer-go/wasmer"
)

var instance *wasmer.Instance
var relativeDir string = "lib/go"

func main() {
	wasmBytes, _ := ioutil.ReadFile(fmt.Sprintf("%s/smartcore_wasi_lib.wasm", relativeDir))
	engine := wasmer.NewEngine()
	store := wasmer.NewStore(engine)

	// Compiles the module
	module, _ := wasmer.NewModule(store, wasmBytes)

	wasiEnv, _ := wasmer.NewWasiStateBuilder("smartcore-wasi-go").
		PreopenDirectory(".").
		Finalize()

	importObject, _ := wasiEnv.GenerateImportObject(store, module)
	newInstance, _ := wasmer.NewInstance(module, importObject)

	instance = newInstance
	init, _ := instance.Exports.GetFunction("init")
	filePtr, _ := getStringPointer(fmt.Sprintf("%s/iris_knn.model", relativeDir), instance)
	init(filePtr)

	loadModel, _ := instance.Exports.GetFunction("load_model")
	startTime := time.Now().UTC().String()
	performances := []int64{}
	numberOfExecutions := 1000
	if os.Getenv("noe") != "" {
		numberOfExecutions, _ = strconv.Atoi(os.Getenv("noe"))
	}
	println("Executing " + strconv.Itoa(numberOfExecutions) + " times")
	for i := 0; i < numberOfExecutions; i++ {
		t0 := time.Now()
		loadModel()
		performances = append(performances, time.Since(t0).Nanoseconds())
	}
	endTime := time.Now().UTC().String()
	stringData := strings.Join(strings.Fields(fmt.Sprint(performances)), ",")
	content := fmt.Sprintf("startTime:%s\nendTime:%s\ndata:%s", startTime, endTime, stringData)
	ioutil.WriteFile("data-go-wasmer.csv", []byte(content), 0644)
}

func getStringPointer(data string, instance *wasmer.Instance) (int32, int) {
	// Set the subject to greet.
	dataPrepared := []byte(data)
	lengthOfSubject := len(dataPrepared) + 1

	// Allocate memory for the subject, and get a pointer to it.
	allocate, _ := instance.Exports.GetFunction("allocate")
	allocateResult, _ := allocate(lengthOfSubject)
	ptr := allocateResult.(int32)

	// Write the subject into the memory.
	memory, _ := instance.Exports.GetMemory("memory")
	memoryData := memory.Data()[ptr:]

	for nth := 0; nth < lengthOfSubject-1; nth++ {
		memoryData[nth] = dataPrepared[nth]
	}

	// C-string terminates by NULL.
	memoryData[lengthOfSubject-1] = 0

	return ptr, lengthOfSubject
}
